pallets/connectors/src/lib.rs

```diff
@@ -0,0 +1,466 @@
+			message: MessageOf<T>,
+			domain: Domain,
+		},
+
+		/// The Router for a given domain was set
+		SetDomainRouter {
+			domain: Domain,
+			router: Router<CurrencyIdOf<T>>,
+		},
+	}
+
+	#[pallet::storage]
+	pub(crate) type DomainRouter<T: Config> =
+		StorageMap<_, Blake2_128Concat, Domain, Router<CurrencyIdOf<T>>>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// A pool could not be found
+		PoolNotFound,
+		/// A tranche could not be found
+		TrancheNotFound,
+		/// Token price is not set
+		MissingPrice,
+		/// Router not set for a given domain
+		MissingRouter,
+		/// Transfer amount must be non-zero
+		InvalidTransferAmount,
+		/// A transfer to a non-whitelisted destination was attempted
+		UnauthorizedTransfer,
+		/// Failed to build Ethereum_Xcm call
+		FailedToBuildEthereumXcmCall,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Set a Domain's router
+		#[pallet::weight(< T as Config >::WeightInfo::set_domain_router())]
+		pub fn set_domain_router(
+			origin: OriginFor<T>,
+			domain: Domain,
+			router: Router<CurrencyIdOf<T>>,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin.clone())?;
+
+			<DomainRouter<T>>::insert(domain.clone(), router.clone());
+			Self::deposit_event(Event::SetDomainRouter { domain, router });
+
+			Ok(())
+		}
+
+		/// Add a pool to a given domain
+		#[pallet::weight(< T as Config >::WeightInfo::add_pool())]
+		pub fn add_pool(
+			origin: OriginFor<T>,
+			pool_id: PoolIdOf<T>,
+			domain: Domain,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+
+			ensure!(
+				T::PoolInspect::pool_exists(pool_id),
+				Error::<T>::PoolNotFound
+			);
+
+			Self::do_send_message(who, Message::AddPool { pool_id }, domain)?;
+
+			Ok(())
+		}
+
+		/// Add a tranche to a given domain
+		#[pallet::weight(< T as Config >::WeightInfo::add_tranche())]
+		pub fn add_tranche(
+			origin: OriginFor<T>,
+			pool_id: PoolIdOf<T>,
+			tranche_id: TrancheIdOf<T>,
+			domain: Domain,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+
+			ensure!(
+				T::PoolInspect::tranche_exists(pool_id, tranche_id),
+				Error::<T>::TrancheNotFound
+			);
+
+			// Send the message through the router
+			//
+			// TODO: retrieve token name and symbol from asset-registry
+			// Depends on https://github.com/centrifuge/centrifuge-chain/issues/842
+			//
+			// TODO: only allow calling add_tranche when
+			// both the name and symbol are non-zero values.
+			Self::do_send_message(
+				who,
+				Message::AddTranche {
+					pool_id,
+					tranche_id,
+					token_name: [0; 32],
+					token_symbol: [0; 32],
+				},
+				domain,
+			)?;
+
+			Ok(())
+		}
+
+		/// Update a token price
+		#[pallet::weight(< T as Config >::WeightInfo::update_token_price())]
+		pub fn update_token_price(
+			origin: OriginFor<T>,
+			pool_id: PoolIdOf<T>,
+			tranche_id: TrancheIdOf<T>,
+			domain: Domain,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+
+			ensure!(
```

Comment:
Hm yeah, if we do that we should change the error `MissingPrice` to something more general, otherwise it will sound like only the price of an non-existing tranche is missing. wdyt?

Reply:
Yeah that makes sense, not entirely sure what the ideal naming is though. Any suggestions?

Reply:
So, we need to check whether the tranche token exists in `add_tranche` whilst not needing to fetch the tranche price. This means that we need to have an error variant for `TrancheNotFound` for this case. 

So I just renamed `MissingPrice` to `MissingTranchePrice` to be more specific it's Tranche-related and left a richer doc comment on what this error can mean.